app/javascript/components/dropdowns/Dropdown.tsx

```diff
@@ -20,123 +19,30 @@ export const Dropdown = ({
-  const [focusIndex, setFocusIndex] = useState<number | null | undefined>()
-  const menuItemElementsRef = useRef<HTMLLIElement[]>([])
-  const id = useMemo(() => {
-    return uuidv4()
-  }, [])
-
-  const handleButtonKeyDown = (e: React.KeyboardEvent) => {
-    switch (e.key) {
-      case 'ArrowDown':
-        e.preventDefault()
-        setOpen(true)
-        setFocusIndex(0)
-
-        break
-      case 'ArrowUp':
-        e.preventDefault()
-        setOpen(true)
-        setFocusIndex(menuItems.length - 1)
-
-        break
-    }
-  }
-
-  const handleItemKeyDown = (e: React.KeyboardEvent, index: number) => {
-    switch (e.key) {
-      case 'ArrowDown':
-        e.preventDefault()
-        setFocusIndex((index + menuItems.length + 1) % menuItems.length)
-
-        break
-      case 'ArrowUp':
-        e.preventDefault()
-        setFocusIndex((index + menuItems.length - 1) % menuItems.length)
-
-        break
-      case 'Tab':
-        setOpen(false)
-
-        break
-      case 'Escape':
-        e.preventDefault()
-        setOpen(false)
-        setFocusIndex(null)
-
-        break
-      case ' ':
-      case 'Enter': {
-        e.preventDefault()
-
-        setOpen(false)
-
-        const link = menuItemElementsRef.current[index]?.querySelector('a')
-        link?.click()
-
-        break
-      }
-    }
-  }
-
-  const handleMenuItemMount = (
-    instance: HTMLLIElement | null,
-    index: number
-  ) => {
-    if (!instance) {
-      return
-    }
-
-    menuItemElementsRef.current[index] = instance
-  }
-
-  useEffect(() => {
-    if (focusIndex === undefined) {
-      return
-    }
-
-    if (focusIndex === null) {
-      buttonRef.current?.focus()
-
-      return
-    }
-
-    menuItemElementsRef.current[focusIndex].focus()
-  }, [open, focusIndex, buttonRef])
+  const {
+    buttonAttributes,
+    panelAttributes,
+    listAttributes,
+    itemAttributes,
+  } = useDropdown(menuItems.length)
 
   return (
     <React.Fragment>
       <button
         className={`${menuButton.className}`}
-        aria-controls={id}
-        aria-haspopup
-        aria-label={menuButton.label}
-        aria-expanded={open ? true : undefined}
         dangerouslySetInnerHTML={{ __html: menuButton.html }}
-        ref={buttonRef}
-        onKeyDown={handleButtonKeyDown}
-        onClick={() => {
-          setOpen(!open)
-        }}
+        aria-label={menuButton.label}
+        {...buttonAttributes}
       />
-      <div ref={panelRef} style={styles.popper} {...attributes.popper}>
-        <ul
-          className={`${menuButton.className}-dropdown`}
-          id={id}
-          role="menu"
-          hidden={!open}
-        >
+      <div {...panelAttributes}>
+        <ul className={`${menuButton.className}-dropdown`} {...listAttributes}>
           {menuItems.map((item, i) => {
             return (
               <li
-                ref={(instance) => handleMenuItemMount(instance, i)}
                 key={item.html}
```

Comment:
I don't know much React, but would setting the key to the index result in less data in the HTML?